easybuild/framework/easyconfig/types.py

```diff
@@ -420,20 +420,39 @@ def to_dependencies(dep_list):
 
 def to_checksums(checksums):
     """Ensure correct element types for list of checksums: convert list elements to tuples."""
-    res = []
-    for checksum in checksums:
-        # each list entry can be:
-        # * a string (MD5 checksum)
-        # * a tuple with 2 elements: checksum type + checksum value
-        # * a list of checksums (i.e. multiple checksums for a single file)
-        if isinstance(checksum, basestring):
-            res.append(checksum)
-        elif isinstance(checksum, (list, tuple)):
-            # 2 elements + only string/int values => a checksum tuple
-            if len(checksum) == 2 and all(isinstance(x, (basestring, int)) for x in checksum):
-                res.append(tuple(checksum))
-            else:
-                res.append(to_checksums(checksum))
+    if isinstance(checksums, (list, tuple)):
+        res = []
+        for checksum in checksums:
+            # each list entry can be:
+            # * a string (MD5 checksum)
+            # * a tuple with 2 elements: checksum type + checksum value
+            # * a list of checksums (i.e. multiple checksums for a single file)
+            if isinstance(checksum, basestring):
+                res.append(checksum)
+            elif isinstance(checksum, (list, tuple)):
+                # 2 elements + only string/int values => a checksum tuple
+                if len(checksum) == 2 and all(isinstance(x, (basestring, int)) for x in checksum):
+                    res.append(tuple(checksum))
+                else:
+                    res.append(to_checksums(checksum))
+
+    elif isinstance(checksums, dict):
```

Comment:
This check should be *in* the for loop, `checksums` is the overarching list of checksums, you intend to add support for `dict` values *in* `checksums`, right?

Reply:
Actually, I had intended to let the top-level checksums variable be a dict... since that's the only way that you can have e.g. more individual checksums than source files and patches...

Reply:
I've not answered the previous comments, since this one seems crucial to agree first!

Reply:
@edmondac This is what I had in mind:

```python
sources = [
    'foo-%(version)s.tar.gz',
    'example-%(version)s-%(arch)s.tar.gz',
]
patches = ['example.patch']
checksums = [
    'def47ee13a30a7698718c4f9ee1569406c8135ffb7febec5158c015a0b0b7590',  # foo-1.2.3.tar.gz
    {
        'example-1.2.3-x86_64.tar.gz': 'f0d5f7cd6137637ef9065431bebeb14333ef13018458135c74024c16161ca1ec',
        'example-1.2.3-ppc64le.tar.gz': '5f24fd6084268f5ffea0efd2027fb4319ba7a6bd3c3659bca1d283cba06b2864',
    },
    '7ce1913278203c45a667df9b7cb6192f46d2a6516fa6e1c1e269a27473d8413c',  # example.patch
]
```

Here, the 2nd item in `checksums` is a dict that holds two checksums for the 2nd source file, depending on the `%(arch)s` value.

What you had in mind is probably this:

```python
checksums = {
    'foo-1.2.3.tar.gz': 'def47ee13a30a7698718c4f9ee1569406c8135ffb7febec5158c015a0b0b7590',
    'example-1.2.3-x86_64.tar.gz': 'f0d5f7cd6137637ef9065431bebeb14333ef13018458135c74024c16161ca1ec',
    'example-1.2.3-ppc64le.tar.gz': '5f24fd6084268f5ffea0efd2027fb4319ba7a6bd3c3659bca1d283cba06b2864',
    'example.patch': '7ce1913278203c45a667df9b7cb6192f46d2a6516fa6e1c1e269a27473d8413c',
}
```

That would work too, but it's more verbose since the filenames of *all* sources/patches need to be mentioned again, even if there's only one checksum for them.

There's another option, which is to extend the "alternative" format we already support for `sources` to also be able to list `checksums` directly together with the `sources` (see https://easybuild.readthedocs.io/en/latest/Writing_easyconfig_files.html#common-easyconfig-param-sources-alt):

```python
sources = [
    'foo-%(version)s.tar.gz',
    {
        'filename': 'example-%(version)s-%(arch)s.tar.gz',
        'checksums': {
            'example-1.2.3-x86_64.tar.gz': 'f0d5f7cd6137637ef9065431bebeb14333ef13018458135c74024c16161ca1ec',
            'example-1.2.3-ppc64le.tar.gz': '5f24fd6084268f5ffea0efd2027fb4319ba7a6bd3c3659bca1d283cba06b2864',
        },
    },
]
```

I think both the ones I had in mind make sense (it comes down to personal preference I guess), and they're probably relatively equally easy to add support for.

Reply:
Ahah - got it. Yes, I can do the first one. I'll probably make a fresh branch for it next week...